src/notebooks/controllers/controllerRegistration.ts

```diff
@@ -161,6 +180,14 @@ export class ControllerRegistration implements IControllerRegistration {
         return this.registeredControllers.get(id);
     }
 
+    private isFiltered(metadata: KernelConnectionMetadata): boolean {
+        const userFiltered = this.kernelFilter.isKernelHidden(metadata);
+        const connectionTypeFiltered = isLocalConnection(metadata) !== this.isLocalLaunch;
+        const urlFiltered =
+            isRemoteConnection(metadata) && !this.serverUriStorage.currentUri?.includes(metadata.baseUrl);
```

Comment:
I think you can use `serverId` instead of `baseUrl` that would still allow you to keep all of this synchronous

Reply:
serverId is a hash of the url? Not sure about that. It's not cryptographically secure. I mean if that was allowed, we wouldn't need encrypted storage.

Reply:
Going to try doing this with async instead.

Reply:
> The difficult part is an update after onDidChangeUri. Maybe if I pass the URI there and during add, I can skip the sync 'currentUri'

See comments above, the previous class `ServerConnectionType._isLocalLaunch`